### PR TITLE
BUG+DOC: Recent Styler Enhancements

### PR DESCRIPTION
**Pull request by attack68**
**DOC**: this edits the `style.ipynb` file to include examples of the 3 recent enhancements to `Styler`:

 - Tooltips
 - External CSS classes to datacells
 - Row and Column styling via table styles

It also addresses 3 minor bugs that I detected when wirting the documentation:

**BUG1** (lines 696-703): `set_td_classes` only attached classes to 'diagonal elements' of the DataFrame due a misaligned comprehension loop:

```
{
r: {
   c: [str(classes.iloc[r, c])] 
   for c, cn in enumerate(classes.columns) if not mask.iloc[r, c]
   }
for r, rn in enumerate(classes.index)
}
```
replaces:
```
{
r: {
    c: [str(classes.iloc[r, c])]
    }
for r, rn in enumerate(classes.index)
for c, cn in enumerate(classes.columns)
if not mask.iloc[r, c]
}
```

**BUG2** (lines 513 to 537): external css class names from the set_td_classes were appended to the HTML id tag, as well as the HTML class tag, due to a premature variable update relating to the `row_dict`. The solution was to re-order so that 'id' tags were generated first and then 'class' was updated with information from the new methods.

**BUG3** (lines 1920-1925): Tooltips were generated for the wrong indexes/columns, since the null rows/columns were dropped and then the interger locations of rows/columns was mapped. The solution was to avoid dropping null rows/columns.

### CODE REVIEW

File: doc/source/user_guide/style.ipynb [hunk 1]
```diff
@@ -61,21 +50,23 @@
    "outputs": [],
    "source": [
     "import pandas as pd\n",
+    "from pandas.io.formats.style import Styler\n",
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:53]: can you move this to the subclassing section / where its actually used (as its not normally needed for most operations)

File: doc/source/user_guide/style.ipynb [hunk 2]
```diff
@@ -102,16 +117,110 @@
...
+    "Table styles are also used to control features which can apply to the whole table at once such as greating a generic hover functionality:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "s = Styler(df, cell_ids=False, uuid_len=1)\n",
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:175]: eg. i think this is the first time you use Styler, so import here

File: doc/source/user_guide/style.ipynb [hunk 3]
```diff
@@ -102,16 +117,110 @@
...
+    "s.set_table_styles({\n",
+    "    'A': [{'selector': '',\n",
+    "           'props': [('color', 'red')]}],\n",
+    "    'B': [{'selector': 'td',\n",
+    "           'props': [('color', 'blue')]}]\n",
+    "}, axis=0, overwrite=False)\n",
+    "s.set_table_styles({\n",
+    "    3: [{'selector': 'td',\n",
+    "           'props': [('color', 'green')]}]\n",
+    "}, axis=1, overwrite=False)"
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:187]: can you render df with a df.style before and then after you have provided this style update

File: doc/source/user_guide/style.ipynb [hunk 4]
```diff
@@ -102,16 +117,110 @@
...
+    "### Styler Functions\n",
+    "\n",
+    "Thirdly we can use the method to pass your style functions into one of the following methods:\n",
     "\n",
-    "When writing style functions, you take care of producing the CSS attribute / value pairs you want. Pandas matches those up with the CSS classes that identify each cell."
+    "- ``Styler.applymap``: elementwise\n",
+    "- ``Styler.apply``: column-/row-/table-wise\n",
+    "\n",
+    "Both of those methods take a function (and some other keyword arguments) and applies your function to the DataFrame in a certain way.\n",
+    "`Styler.applymap` works through the DataFrame elementwise.\n",
```
Comment (jreback) [inline doc/source/user_guide/style.ipynb:211]: not sure how easy this is to link to the actual docs for these but would be great if possible.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some comments

cc @TomAugspurger

**Comment by jreback** [comment]
@attack68 can you also add tests for the bugs you have fixed (e.g. simply a check on the generated html might be fine), just to lock down these cases.

File: doc/source/user_guide/style.ipynb (hunk 4 above)
Comment (attack68) [reply doc/source/user_guide/style.ipynb:289]: i tried. i think it is possible but the relative path structure is quite complicated and i couldn't get it to work. never done it before so I have left out.

**Comment by attack68** [comment]
@jreback think this is ready now..

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68 

pls check the rendered docs when they build: https://pandas.pydata.org/docs/dev/

**Pull request merged**